scipy/stats/tests/test_sampling.py

```diff
@@ -0,0 +1,380 @@
+            (lambda x: x * x, r"51 : hat\(x\) < PDF\(x\)"),
+            (lambda x: None, r"must be real number, not NoneType"),
+            (lambda x: np.inf, r"50 : PDF\(x\) overflow"),
+            (lambda x: np.nan, r"51 : cannot create bounded hat!"),
+            (
+                lambda x: x * x / 2 + 2,
+                r"51 : dTfx0 < dTfx1 \(x0<x1\). PDF not T-concave!",
+            ),
+            (lambda: 1.0, r"takes 0 positional arguments but 1 was given")
+        ],
+    )
+    @pytest.mark.parametrize(
+        "kwargs", [{"dpdf": (lambda x: x), "domain": (1, 10)}]
+    )
+    def test_bad_pdf(self, bad_pdf, msg, kwargs):
+        with pytest.raises(Exception, match=msg):
+            TransformedDensityRejection(bad_pdf, **kwargs)
+
+    @pytest.mark.parametrize(
+        "bad_dpdf, msg",
+        [
+            (lambda x: foo, r"name 'foo' is not defined"),
+            (lambda x: x * x, r"51 : hat\(x\) < PDF\(x\)"),
+            (lambda x: None, r"must be real number, not NoneType"),
+            (lambda x: np.inf, r"51 : cannot create bounded hat!"),
+            (lambda x: np.nan, r"51 : cannot create bounded hat!"),
+            (lambda: 1.0, r"takes 0 positional arguments but 1 was given")
+        ],
+    )
+    @pytest.mark.parametrize(
+        "kwargs", [{"pdf": (lambda x: x * x / 2), "domain": (1, 10)}]
+    )
+    def test_bad_dpdf(self, bad_dpdf, msg, kwargs):
+        with pytest.raises(Exception, match=msg):
+            TransformedDensityRejection(dpdf=bad_dpdf, **kwargs)
+
+    @pytest.mark.parametrize(
+        "domain", [(0, 0), (1, 0), (np.inf, np.inf), (-np.inf, -np.inf)]
+    )
+    def test_bad_domain(self, domain):
+        with pytest.raises(RuntimeError, match=r"left >= right"):
+            TransformedDensityRejection(
+                lambda x: x, lambda x: 1, domain=domain
+            )
+
+    def test_bad_sized_domain(self):
+        with pytest.raises(ValueError, match=r"must be a length 2 tuple"):
+            TransformedDensityRejection(
+                lambda x: x, lambda x: 1, domain=(1, 2, 3)
+            )
+
+    @pytest.mark.parametrize(
+        "domain",
+        [
+            (np.nan, np.nan),
+            (np.inf, np.nan),
+            (np.nan, -np.inf),
+            (np.nan, 0),
+            (-1, np.nan),
+            (0, float("nan")),
+        ],
+    )
+    def test_nan_domain(self, domain):
+        with pytest.raises(ValueError, match=r"only non-nan values"):
+            TransformedDensityRejection(
+                lambda x: x, lambda x: 1, domain=domain
+            )
+
+    def test_bad_cpoints(self):
+        # test bad cpoints
+        with pytest.warns(
+            UserWarning, match=r"number of starting points < 0"
+        ):
+            TransformedDensityRejection(
+                lambda x: x, lambda x: 1, domain=(0, 10), cpoints=-10
+            )
+
+        with pytest.warns(UserWarning, match=r"hat/squeeze ratio too small"):
```

Comment:
> how did you come up with this test case? could you add a comment why a warning is raised?

I think it was just an intuitive test case i.e. Do I run into a problem when `cpoints` to too small? I just tested it and ran into some warnings which I have tested here. I will add a comment on that.

Reply:
I have removed this test case because I was getting confused while discerning the reason for adding this. I have now added comments for each test case. So, I think it should be easier to review them now.